src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,243 @@
+            in_flight: inner.in_flight as u64
+        });
+
+        let rtt = inner.current_rtt.update(rtt);
+        let avg = inner.past_rtt.average();
+
+        if avg == 0.0 {
+            // No past measurements, set up initial values.
+            inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(rtt);
+        } else if avg > 0.0 && now >= inner.next_update {
+            emit!(AutoConcurrencyAveragedRtt {
+                rtt: Duration::from_secs_f64(rtt)
+            });
+
+            #[cfg(test)]
+            {
+                stats.averaged_rtt.add(rtt, now);
+                stats.concurrency_limit.add(inner.current_limit, now);
+            }
+
+            let threshold = avg * THRESHOLD_RATIO;
+
+            // Back pressure responses, either explicit or implicit due
+            // to increasing response times, trigger a decrease in the
+            // concurrency limit.
+            if inner.current_limit > 1 && (inner.had_back_pressure || rtt >= avg + threshold) {
+                // Decrease (multiplicative) the current concurrency limit
+                let to_forget = inner.current_limit / 2;
+                self.semaphore.forget_permits(to_forget);
+                inner.current_limit -= to_forget;
+            }
+            // Normal quick responses trigger an increase in the
+            // concurrency limit. Note that we only check this if we had
+            // requests to go beyond the current limit to prevent
+            // increasing the limit beyond what we have evidence for.
+            else if inner.current_limit < self.max
+                && inner.reached_limit
+                && !inner.had_back_pressure
+                && rtt <= avg
+            {
+                // Increase (additive) the current concurrency limit
+                self.semaphore.add_permits(1);
+                inner.current_limit += 1;
+            }
+            emit!(AutoConcurrencyLimit {
+                concurrency: inner.current_limit as u64,
+            });
+
+            // Reset values for next interval
+            let new_avg = inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(new_avg);
+            inner.current_rtt.reset();
+            inner.had_back_pressure = false;
+            inner.reached_limit = false;
+        }
+    }
+}
+
+impl<L> Controller<L>
+where
+    L: RetryLogic,
+{
+    pub(super) fn adjust_to_response(
+        &self,
+        start: Instant,
+        response: &Result<L::Response, crate::Error>,
+    ) {
+        let is_back_pressure = match response {
+            Ok(_) => false,
+            Err(err) => {
+                if let Some(err) = err.downcast_ref::<L::Error>() {
+                    self.logic.is_retriable_error(err)
+                } else if err.downcast_ref::<Elapsed>().is_some() {
+                    true
+                } else {
+                    panic!("Unhandled error response! {:?}", err)
```

Comment:
Do we have anything that will prevent this panic at compile time? I don't want to get into a situation where we're panicking because we forgot to implement a certain error type.

Reply:
No, unfortunatly there is nothing we _can_ do at compile time to prevent this panic. This downcast is only required because the error type of the `Timeout` layer is `Box<Error>` instead of an enum. The only errors that `Timeout` produces is either the input error type or `Elapsed`, so it should not be possible to trigger the panic, but I can't guarantee that.

I did consider rewriting the `Timeout` layer as well so that it's `type Error` was an actual enum instead of a `Box`, but figured that was out of scope for this as well. I'm not happy with this solution.

Reply:
> I did consider rewriting the Timeout layer as well so that it's type Error was an actual enum instead of a Box, but figured that was out of scope for this as well. I'm not happy with this solution.

Should we open this as a tech-debt issue?

Reply:
I would, but it's a tech-debt issue in the `tower` crate, which would also introduce a breaking API change. How does that factor in?

Reply:
Ah. So it's unlikely to be accepted? If that's the case it sounds like it's not worth it, but I am not a good judge of the benefit it would unlock.

Reply:
I would guess it to be unlikely to be accepted, yes, but you never know. The benefits are avoiding a needless `Box` and enforcing type-correctness on users of the timeout layer error type, but I don't know how the tower crew would perceive those benefits.

Reply:
It sounds like it's worth opening an issue on that repo first to see if the change would be accepted. If you agree, would you mind doing that and referencing it here?

Reply:
Will do.

Reply:
Opened https://github.com/tower-rs/tower/issues/465

Reply:
I wonder if we could do something tricky where we bundle the inner service, retry logic, this layer, and timeout layer into one struct (or even constructor fn) where we can bound the inner `Service::Error` to the `RetryLogic::Error`. That wouldn't get rid of the box in the middle, but it'd help prevent us from constructing a mismatched stack.

Reply:
I'm sorry, this isn't tricky enough? :astonished:

Reply:
Just brainstorming :smile: 

If you're comfortable with the current risk level, that's good enough for me!

Reply:
It turns out tower used to do exactly what we want, but it had issues too, particularly when layers were nested. See https://github.com/tower-rs/tower/issues/131

Reply:
Based on the discussion above, it seems like this can be changed to `unreachable!("unhandled error response! {:?}", err)` to better capture the intent of the branch.

Reply:
Agreed.